extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -428,32 +462,67 @@ export class AmpAutocomplete extends AMP.BaseElement {
     // Client-side filtering.
     input = input.toLocaleLowerCase();
     const itemsExpr = this.element.getAttribute('filter-value') || 'value';
-    const filteredData = data.filter(item => {
-      if (typeof item === 'object') {
-        item = getValueForExpr(/** @type {!JsonObject} */(item), itemsExpr);
-      }
-      userAssert(typeof item === 'string',
-          `${TAG} data property "${itemsExpr}" must map to string type.`);
-      item = item.toLocaleLowerCase();
-      switch (this.filter_) {
-        case FilterType.SUBSTRING:
-          return includes(item, input);
-        case FilterType.PREFIX:
-          return startsWith(item, input);
-        case FilterType.TOKEN_PREFIX:
-          return this.tokenPrefixMatch_(item, input);
-        case FilterType.FUZZY:
-          throw new Error(`Filter not yet supported: ${this.filter_}`);
-        case FilterType.CUSTOM:
-          throw new Error(`Filter not yet supported: ${this.filter_}`);
-        default:
-          throw new Error(`Unexpected filter: ${this.filter_}`);
+    const filteredData = this.copyJsonObject_(data.filter(item => {
+      return this.filterMatch_(item, input, itemsExpr);
+    }));
+    filteredData.map(item => {
+      const category = getValueForExpr(/**@type {!JsonObject} */(item), 'category');
+      if (category) {
+        const opts = getValueForExpr(/**@type {!JsonObject} */(item), 'items');
+        const filteredOpts = opts.filter(item => {
+          return this.filterMatch_(item, input, itemsExpr);
+        });
+        item['items'] = filteredOpts;
       }
+      return item;
     });
 
     return this.truncateToMaxEntries_(filteredData);
   }
 
+  /**
+   * Client-side filtering method.
+   *
+   * Returns true if the given item is a match on the given input according
+   * to the "filter" attribute on this amp-autocomplete instance.
+   *
+   * @param {!Object|string} item
+   * @param {string} input
+   * @param {string} itemsExpr
+   * @return {boolean}
+   * @private
+   */
+  filterMatch_(item, input, itemsExpr) {
+    if (typeof item === 'object') {
+      const category = getValueForExpr(/**@type {!JsonObject} */(item), 'category');
```

Comment:
This works, but why not use `item['category']`?

Reply:
Good point. I was actually wondering this myself but remembered that I used it in a different situation for reasons that aren't applicable here.